raiden/network/proxies/token_network.py

```diff
@@ -1177,66 +1157,164 @@ def update_transfer(
+                        f'value expected={nonce} actual={partner_details.nonce}'
+                    )
+                    raise RaidenUnrecoverableError(msg)
+
+                if channel_data.state < ChannelState.CLOSED:
+                    msg = (
+                        f'The channel state changed unexpectedly. '
+                        f'block=({mining_block}) onchain_state={channel_data.state}'
+                    )
+                    raise RaidenUnrecoverableError(msg)
+
+                raise RaidenUnrecoverableError('update transfer failed for an unknown reason')
+
+        else:
+            # The transaction would have failed if sent, figure out why.
+
+            # The latest block can not be used reliably because of reorgs,
+            # therefore every call using this block has to handle pruned data.
+            failed_at = self.proxy.jsonrpc_client.get_block('latest')
+            failed_at_blockhash = encode_hex(failed_at['hash'])
+            failed_at_blocknumber = failed_at['number']
+
+            # This check contains a race condition, it could be the case that a
+            # new block is mined changing the account's balance.
+            # https://github.com/raiden-network/raiden/issues/3890#issuecomment-485857726
             self.proxy.jsonrpc_client.check_for_insufficient_eth(
                 transaction_name='updateNonClosingBalanceProof',
-                transaction_executed=transaction_executed,
+                transaction_executed=False,
                 required_gas=GAS_REQUIRED_FOR_UPDATE_BALANCE_PROOF,
-                block_identifier=block,
+                block_identifier=failed_at_blocknumber,
             )
+
             detail = self._detail_channel(
                 participant1=self.node_address,
                 participant2=partner,
-                block_identifier=block,
+                block_identifier=failed_at_blockhash,
                 channel_identifier=channel_identifier,
             )
-            msg: Optional[str]
-            error_type: ErrorType
-            if detail.settle_block_number < to_compare_block:
-                error_type = RaidenRecoverableError
+
+            if detail.state < ChannelState.CLOSED:
                 msg = (
-                    'updateNonClosingBalanceProof transaction '
-                    'was mined after settlement finished'
+                    f'cannot call update_transfer channel has not been closed yet. '
+                    f'current_state={detail.state}'
                 )
-            else:
-                msg = self._check_channel_state_for_update(
-                    channel_identifier=channel_identifier,
-                    closer=partner,
-                    update_nonce=nonce,
-                    block_identifier=block,
+                raise RaidenUnrecoverableError(msg)
+
+            if detail.state > ChannelState.CLOSED:
```

Comment:
humm, maybe this can be improved by using `>= SETTLED`

Reply:
Note, this check here is different from the one [on the precondition](https://github.com/raiden-network/raiden/pull/3897/files#diff-d02574a7a7eca826412780e80a6326a3R1108). This is because the precondition must always be valid at the given block, where as this can race on-chain, so later states are recoverable errors instead of unrecoverable.